tests/ui/macros/trace_faulty_macros.stderr

```diff
@@ -69,6 +72,23 @@ LL | #[derive(Debug)]
 LL | fn use_derive_macro_as_attr() {}
    | -------------------------------- not a `struct`, `enum` or `union`
 
+error: expected expression, found `1 + 1`
+  --> $DIR/trace_faulty_macros.rs:49:37
+   |
+LL |     (let $p:pat = $e:expr) => {test!(($p,$e))};
+   |                                          -- this is interpreted as pattern `1 + 1` (in expansion #2)
+...
+LL |     (($p:pat, $e:pat)) => {let $p = $e;};
+   |                                     ^^ expected expression
+...
+LL |     test!(let x = 1+1);
+   |     ------------------
+   |     |             |
+   |     |             this is interpreted as expression `1 + 1` (in expansion #1)
+   |     in this macro invocation
+   |
+   = note: this error originates in the macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Comment:
I know, ideally we would actually point at `$e:pat`, but without this context it is even harder to understand that the reason `1 + 1` is understood as a pattern is that there's multiple levels of macro evaluation at play.

Reply:
#71039 is two problems:
1. reinterpreting expr non-terminals as patterns leads to confusing errors with multiple layers of nesting,
2. unrelated to nesting, using pattern non-terminals in expr position itself leads to weird errors:


this is an example of (2.), which would be fixed if we were to point at the `:pat` kind, and which (afaict) isn't helped by this pr?
```
macro_rules! m {
    ($e:pat) => { let x = $e; }
    //~^ ERROR expected expression, found `1`
}

fn main() {
    m!(1);
}
```

Reply:
> which (afaict) isn't helped by this pr?

maybe this is a mischaracterization -- the context added by this pr might help a bit, but ideally the only thing we'd need to add a new label for is the `:pat` part.

Reply:
Yeah, I think I figured out a way to add the span for `e:pat` as the `def_site` context of the span for `$e`, but I'll need to talk with petrochenkov to see if that's reasonable. If we had that, it could also be used in a bunch of other errors as well.